test/services/mock_api.test.ts

```diff
@@ -0,0 +1,135 @@
+import chai, { expect } from "chai";
+import chaiAsPromised from "chai-as-promised";
+import { BigNumber } from "ethers";
+
+import { OrderKind } from "../../src/ts";
+
+import { MockApi } from "./mock_api";
+
+chai.use(chaiAsPromised);
+
+describe("MockApi", () => {
+  let api = new MockApi();
+
+  const feeInput = {
+    sellToken: "0x" + "42".repeat(20),
+    buyToken: "0x" + "21".repeat(20),
+    kind: OrderKind.SELL,
+    amount: 31337,
+  };
+  const feeOutput = BigNumber.from(42);
+  const feeError = {
+    errorType: "MockError",
+    description: "a mocked error",
+  };
+
+  beforeEach(() => {
+    api = new MockApi();
+  });
+
+  it("mocks successful api calls", async () => {
+    api.mock.getFee.withArgs(feeInput).returns(feeOutput);
+    await expect(api.getFee(feeInput)).to.eventually.deep.equal(feeOutput);
+  });
+
+  describe("mocks failing api calls", () => {
+    it("with error", async () => {
+      api.mock.getFee.withArgs(feeInput).throwsWith(feeError);
+      await expect(api.getFee(feeInput))
+        .to.eventually.be.rejected.and.be.an.instanceOf(Error)
+        .and.have.deep.property("apiError", feeError);
+    });
+
+    it("without error", async () => {
+      api.mock.getFee.withArgs(feeInput).throws();
+      await expect(api.getFee(feeInput))
+        .to.eventually.be.rejected.and.be.an.instanceOf(Error)
+        .and.not.have.property("apiError");
+    });
+  });
+
+  it("throws if no expectations are set", async () => {
+    expect(() => api.getFee(feeInput)).to.throw();
+  });
+
+  describe("cannot set related parameters twice", () => {
+    it("input values", async () => {
+      const builder = api.mock.getFee.withArgs(feeInput);
+      expect(() => builder.withArgs(feeInput)).to.throw();
+    });
+
+    it("output values", async () => {
+      const builder = api.mock.getFee.returns(feeOutput);
+      expect(() => builder.returns(feeOutput)).to.throw();
```

Comment:
Is it be possible to have the mock API return the same response twice when called on the same input? E.g. if we want to test a retry mechanism or the like?